In [32]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import association_rules, apriori

In [33]:
df = pd.read_csv("groceries.csv")
df

0                    1                 2  \
0            citrus fruit  semi-finished bread         margarine   
1          tropical fruit               yogurt            coffee   
2              whole milk                  NaN               NaN   
3               pip fruit               yogurt     cream cheese    
4        other vegetables           whole milk    condensed milk   
...                   ...                  ...               ...   
9830              sausage              chicken              beef   
9831    cooking chocolate                  NaN               NaN   
9832              chicken         citrus fruit  other vegetables   
9833  semi-finished bread        bottled water              soda   
9834              chicken       tropical fruit  other vegetables   

                             3              4               5  \
0                  ready soups            NaN             NaN   
1                          NaN            NaN             NaN   
2                          NaN            NaN             NaN   
3                 meat spreads            NaN             NaN   
4     long life bakery product            NaN             NaN   
...                        ...            ...             ...   
9830            hamburger meat   citrus fruit          grapes   
9831                       NaN            NaN             NaN   
9832                    butter         yogurt  frozen dessert   
9833              bottled beer            NaN             NaN   
9834                   vinegar  shopping bags             NaN   

                    6           7       8                   9  ...   22   23  \
0                 NaN         NaN     NaN                 NaN  ...  NaN  NaN   
1                 NaN         NaN     NaN                 NaN  ...  NaN  NaN   
2                 NaN         NaN     NaN                 NaN  ...  NaN  NaN   
3                 NaN         NaN     NaN                 NaN  ...  NaN  NaN   
4                 NaN         NaN     NaN                 NaN  ...  NaN  NaN   
...               ...         ...     ...                 ...  ...  ...  ...   
9830  root vegetables  whole milk  butter  whipped/sour cream  ...  NaN  NaN   
9831              NaN         NaN     NaN                 NaN  ...  NaN  NaN   
9832    domestic eggs  rolls/buns     rum     cling film/bags  ...  NaN  NaN   
9833              NaN         NaN     NaN                 NaN  ...  NaN  NaN   
9834              NaN         NaN     NaN                 NaN  ...  NaN  NaN   

       24   25   26 Unnamed: 27 Unnamed: 28 Unnamed: 29 Unnamed: 30  \
0     NaN  NaN  NaN         NaN         NaN         NaN         NaN   
1     NaN  NaN  NaN         NaN         NaN         NaN         NaN   
2     NaN  NaN  NaN         NaN         NaN         NaN         NaN   
3     NaN  NaN  NaN         NaN         NaN         NaN         NaN   
4     NaN  NaN  NaN         NaN         NaN         NaN         NaN   
...   ...  ...  ...         ...         ...         ...         ...   
9830  NaN  NaN  NaN         NaN         NaN         NaN         NaN   
9831  NaN  NaN  NaN         NaN         NaN         NaN         NaN   
9832  NaN  NaN  NaN         NaN         NaN         NaN         NaN   
9833  NaN  NaN  NaN         NaN         NaN         NaN         NaN   
9834  NaN  NaN  NaN         NaN         NaN         NaN         NaN   

     Unnamed: 31  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
...          ...  
9830         NaN  
9831         NaN  
9832         NaN  
9833         NaN  
9834         NaN  

[9835 rows x 32 columns]

In [34]:
records=list()

for i in range(0,9835):
     records.append([str(df.values[i,j]) for j in range(0,32)])
records

[['citrus fruit',
  'semi-finished bread',
  'margarine',
  'ready soups',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan'],
 ['tropical fruit',
  'yogurt',
  'coffee',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan'],
 ['whole milk',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan'],
 ['pip fruit',
  'yogurt',
  'cream cheese ',
  'meat spreads',
  'nan

## Preparing the data for Association Rule Mining

In [37]:
te = TransactionEncoder()
arr=te.fit(records).transform(records)
df = pd.DataFrame(arr, columns=te.columns_)



del df['nan']
df.dropna()
df.head(11)
df=df.iloc[1:]
df.head(11)

Instant food products  UHT-milk  abrasive cleaner  artif. sweetener  \
1                   False     False             False             False   
2                   False     False             False             False   
3                   False     False             False             False   
4                   False     False             False             False   
5                   False     False              True             False   
6                   False     False             False             False   
7                   False      True             False             False   
8                   False     False             False             False   
9                   False     False             False             False   
10                  False     False             False             False   
11                  False     False             False             False   

    baby cosmetics  baby food   bags  baking powder  bathroom cleaner   beef  \
1            False      False  False          False             False  False   
2            False      False  False          False             False  False   
3            False      False  False          False             False  False   
4            False      False  False          False             False  False   
5            False      False  False          False             False  False   
6            False      False  False          False             False  False   
7            False      False  False          False             False  False   
8            False      False  False          False             False  False   
9            False      False  False          False             False  False   
10           False      False  False          False             False  False   
11           False      False  False          False             False  False   

    ...  turkey  vinegar  waffles  whipped/sour cream  whisky  white bread  \
1   ...   False    False    False               False   False        False   
2   ...   False    False    False               False   False        False   
3   ...   False    False    False               False   False        False   
4   ...   False    False    False               False   False        False   
5   ...   False    False    False               False   False        False   
6   ...   False    False    False               False   False        False   
7   ...   False    False    False               False   False        False   
8   ...   False    False    False               False   False        False   
9   ...   False    False    False               False   False        False   
10  ...   False    False    False               False   False         True   
11  ...   False    False    False               False   False        False   

    white wine  whole milk  yogurt  zwieback  
1        False       False    True     False  
2        False        True   False     False  
3        False       False    True     False  
4        False        True   False     False  
5        False        True    True     False  
6        False       False   False     False  
7        False       False   False     False  
8        False       False   False     False  
9        False        True   False     False  
10       False       False   False     False  
11       False        True    True     False  

[11 rows x 169 columns]

## 9) No. of transactions and itemsets in the Dataframe

In [15]:
df.shape

(9835, 170)

There are 9835 Transactions and 169 itemsets

## 12) Find top selling items with minimum support of 2%

In [16]:
frequent_itemsets = apriori(df, min_support=0.02, use_colnames=True)
frequent_itemsets['length'] = frequent_itemsets['itemsets'].apply(lambda x: len(x))
frequent_itemsets[ (frequent_itemsets['length'] == 1) & (frequent_itemsets['support'] >= 0.02) ].sort_values(by='support', ascending=False)

support                    itemsets  length
35  0.999898                       (nan)       1
58  0.255516                (whole milk)       1
40  0.193493          (other vegetables)       1
44  0.183935                (rolls/buns)       1
50  0.174377                      (soda)       1
59  0.139502                    (yogurt)       1
5   0.110524             (bottled water)       1
45  0.108998           (root vegetables)       1
54  0.104931            (tropical fruit)       1
48  0.098526             (shopping bags)       1
47  0.093950                   (sausage)       1
41  0.088968                    (pastry)       1
15  0.082766              (citrus fruit)       1
4   0.080529              (bottled beer)       1
37  0.079817                (newspapers)       1
10  0.077682               (canned beer)       1
42  0.075648                 (pip fruit)       1
24  0.072293     (fruit/vegetable juice)       1
56  0.071683        (whipped/sour cream)       1
6   0.064870               (brown bread)       1
20  0.063447             (domestic eggs)       1
21  0.058973               (frankfurter)       1
32  0.058566                 (margarine)       1
16  0.058058                    (coffee)       1
43  0.057651                      (pork)       1
7   0.055414                    (butter)       1
18  0.053279                      (curd)       1
1   0.052466                      (beef)       1
36  0.052364                   (napkins)       1
14  0.049619                 (chocolate)       1
23  0.048094         (frozen vegetables)       1
13  0.042908                   (chicken)       1
57  0.042095               (white bread)       1
17  0.039654             (cream cheese )       1
55  0.038434                   (waffles)       1
46  0.037824               (salty snack)       1
31  0.037417  (long life bakery product)       1
19  0.037112                   (dessert)       1
53  0.033859                     (sugar)       1
0   0.033452                  (UHT-milk)       1
27  0.033249            (hamburger meat)       1
2   0.033249                   (berries)       1
29  0.032944          (hygiene articles)       1
39  0.031012                    (onions)       1
52  0.030402       (specialty chocolate)       1
9   0.029893                     (candy)       1
22  0.028368              (frozen meals)       1
34  0.028368           (misc. beverages)       1
38  0.028063                       (oil)       1
8   0.027961               (butter milk)       1
51  0.027351             (specialty bar)       1
26  0.026029                       (ham)       1
3   0.026029                 (beverages)       1
33  0.025826                      (meat)       1
30  0.025013                 (ice cream)       1
49  0.024504             (sliced cheese)       1
28  0.024504               (hard cheese)       1
11  0.023284                  (cat food)       1
25  0.022369                    (grapes)       1
12  0.021047               (chewing gum)       1

## 13) Find all frequent itemsets with minimum support of 5%

In [7]:
frequent_itemsets[(frequent_itemsets['length'] != 1)&(frequent_itemsets['support'] >= 0.05)].sort_values(by='support', ascending=False)

support                        itemsets  length
91   0.074835  (whole milk, other vegetables)       2
103  0.056634        (rolls/buns, whole milk)       2
119  0.056024            (yogurt, whole milk)       2

## 14) Find all frequent itemsets of length 2 with minimum support of 2%

In [8]:
frequent_itemsets[(frequent_itemsets['length'] == 2)&(frequent_itemsets['support'] >= 0.02) ].sort_values(by='support', ascending=False)

support                             itemsets  length
91   0.074835       (whole milk, other vegetables)       2
103  0.056634             (rolls/buns, whole milk)       2
119  0.056024                 (yogurt, whole milk)       2
106  0.048907        (whole milk, root vegetables)       2
85   0.047382  (other vegetables, root vegetables)       2
..        ...                                  ...     ...
75   0.020539            (whole milk, frankfurter)       2
96   0.020437          (pip fruit, tropical fruit)       2
60   0.020437           (bottled beer, whole milk)       2
76   0.020437      (frozen vegetables, whole milk)       2
67   0.020031           (butter, other vegetables)       2

[61 rows x 3 columns]

## 15) Find the top 10 association ruleswith minimum support of 2%, sorted by confidence in descending order

In [9]:
association_rules(frequent_itemsets, metric="support", min_threshold=0.02).sort_values(by='confidence', ascending=False)

antecedents          consequents  \
129           (yogurt, other vegetables)         (whole milk)   
19                              (butter)         (whole milk)   
27                                (curd)         (whole milk)   
124  (root vegetables, other vegetables)         (whole milk)   
123        (whole milk, root vegetables)   (other vegetables)   
..                                   ...                  ...   
78                          (whole milk)               (pork)   
1                           (whole milk)               (beef)   
32                          (whole milk)        (frankfurter)   
35                          (whole milk)  (frozen vegetables)   
3                           (whole milk)       (bottled beer)   

     antecedent support  consequent support   support  confidence      lift  \
129            0.043416            0.255516  0.022267    0.512881  2.007235   
19             0.055414            0.255516  0.027555    0.497248  1.946053   
27             0.053279            0.255516  0.026131    0.490458  1.919481   
124            0.047382            0.255516  0.023183    0.489270  1.914833   
123            0.048907            0.193493  0.023183    0.474012  2.449770   
..                  ...                 ...       ...         ...       ...   
78             0.255516            0.057651  0.022166    0.086749  1.504719   
1              0.255516            0.052466  0.021251    0.083168  1.585180   
32             0.255516            0.058973  0.020539    0.080382  1.363029   
35             0.255516            0.048094  0.020437    0.079984  1.663094   
3              0.255516            0.080529  0.020437    0.079984  0.993237   

     leverage  conviction  
129  0.011174    1.528340  
19   0.013395    1.480817  
27   0.012517    1.461085  
124  0.011076    1.457687  
123  0.013719    1.533320  
..        ...         ...  
78   0.007435    1.031862  
1    0.007845    1.033487  
32   0.005470    1.023280  
35   0.008149    1.034663  
3   -0.000139    0.999408  

[134 rows x 9 columns]

## 16) Find association rules with minimum support of 2% and lift of more than 1

In [10]:
df=association_rules(frequent_itemsets, metric="support", min_threshold=0.02)
df[df['lift'] > 1 ]


antecedents                     consequents  \
0                            (beef)                    (whole milk)   
1                      (whole milk)                          (beef)   
4                (other vegetables)                 (bottled water)   
5                   (bottled water)              (other vegetables)   
6                      (rolls/buns)                 (bottled water)   
..                              ...                             ...   
129      (yogurt, other vegetables)                    (whole milk)   
130  (whole milk, other vegetables)                        (yogurt)   
131                        (yogurt)  (whole milk, other vegetables)   
132                    (whole milk)      (yogurt, other vegetables)   
133              (other vegetables)            (yogurt, whole milk)   

     antecedent support  consequent support   support  confidence      lift  \
0              0.052466            0.255516  0.021251    0.405039  1.585180   
1              0.255516            0.052466  0.021251    0.083168  1.585180   
4              0.193493            0.110524  0.024809    0.128219  1.160101   
5              0.110524            0.193493  0.024809    0.224471  1.160101   
6              0.183935            0.110524  0.024199    0.131564  1.190373   
..                  ...                 ...       ...         ...       ...   
129            0.043416            0.255516  0.022267    0.512881  2.007235   
130            0.074835            0.139502  0.022267    0.297554  2.132979   
131            0.139502            0.074835  0.022267    0.159621  2.132979   
132            0.255516            0.043416  0.022267    0.087147  2.007235   
133            0.193493            0.056024  0.022267    0.115081  2.054131   

     leverage  conviction  
0    0.007845    1.251315  
1    0.007845    1.033487  
4    0.003424    1.020297  
5    0.003424    1.039945  
6    0.003870    1.024228  
..        ...         ...  
129  0.011174    1.528340  
130  0.011828    1.225003  
131  0.011828    1.100890  
132  0.011174    1.047905  
133  0.011427    1.066737  

[126 rows x 9 columns]